File: Python/multi_threaded.py
import cv2
import numpy as np
import threading
import multiprocessing
from typing import Final

from imageproc_utils import recolor_image

OUTPUT_IMAGE: Final[str] = 'result_image/multi-threaded.jpg'


def multi_threaded(img: np.ndarray) -> None:
    number_of_threads: Final[int] = multiprocessing.cpu_count()
    img_height: Final[int] = img.shape[0] // number_of_threads
    img_width: Final[int] = img.shape[1]

    threads: list[threading.Thread] = []
    for thread_number in range(number_of_threads):
        thread = threading.Thread(
            target=recolor_image,
            args=(
                img,
                img,
                0,
                thread_number * img_height,
                img_width,
                img_height,
            )
        )
        threads.append(thread)

    for thread in threads:
        thread.start()

    for thread in threads:
        thread.join()

    cv2.imwrite(OUTPUT_IMAGE, img)

File: C++/src/multi_threaded.cpp
#include "multi_threaded.hpp"
#include "image_processing.hpp"
#include <opencv2/imgcodecs.hpp>
#include <thread>
#include <vector>

namespace multi_threaded
{

  void processImage(cv::Mat &img)
  {
    int numberOfThreads = std::thread::hardware_concurrency();
    std::vector<std::thread> threads;
    threads.reserve(numberOfThreads);

    int imgHeight = img.rows / numberOfThreads;
    int imgWidth = img.cols;

    for (int i = 0; i < numberOfThreads; ++i)
    {
      threads.emplace_back([&, i]()
                           { color_changer::recolorImage(img, img, 0, i * imgHeight, imgWidth, imgHeight); });
    }

    for (auto &thread : threads)
    {
      thread.join();
    }

    cv::imwrite("../../result_image/multi-threaded.jpg", img);
  }

} // namespace multi_threaded
